feat(query): Spread override  and app based spread support

TanviBhavsar opened the pull request:
**Pull Request checklist**

- [x] The commit(s) message(s) follows the contribution [guidelines](CONTRIBUTING.md) ?
- [x] Tests for the changes have been added (for bug fixes / features) ?
- [x] Docs have been added / updated (for bug fixes / features) ?

**Current behavior :** 
default-spread  from config value is used by FiloDB Cli, FiloDB HTTP and Remote Read. Query Service uses spread override for app  in config


**New behavior :**
Spread can be provided  in query by Filo-cli, FiloDB HTTP and Query service. If spread is not provided, spread. default  or app specific spread from config will be used by QueryActor.

velvia commented on conf/timeseries-dev-source.conf:
```diff
@@ -1,5 +1,5 @@
     dataset = "prometheus"
-    num-shards = 4
+    num-shards = 8
```
Tanvi have you ran through all of the local testing in the README to ensure queries still work?  Please also test the path involving local gateway also.  This change changes a lot of local standalone scenarios and needs to be tested thoroughly

TanviBhavsar replied:
Sorry I checked this in by mistake

velvia commented on coordinator/src/main/scala/filodb.coordinator/QueryActor.scala:
```diff
@@ -55,6 +57,14 @@ final class QueryActor(memStore: MemStore,
 
+  var filodbSpreadMap = new mutable.HashMap[String, Int]()
+  config.getConfig("filodb.spread.override").entrySet().asScala.
+    foreach(x => filodbSpreadMap.put(x.getKey(), x.getValue().unwrapped().asInstanceOf[Int]))
```
Use `foreach { x => .... }` style

TanviBhavsar replied:
Done

velvia commented on coordinator/src/main/scala/filodb.coordinator/QueryActor.scala:
```diff
@@ -91,11 +101,20 @@ final class QueryActor(memStore: MemStore,
     lpRequests.increment
     try {
-      val execPlan = queryEngine2.materialize(q.logicalPlan, q.queryOptions)
+      val execPlan = queryEngine2.materialize(q.logicalPlan,q.queryOptions,getSpreadProvider(q.queryOptions))
```
space after ,

TanviBhavsar replied:
Done

velvia commented on coordinator/src/main/scala/filodb.coordinator/client/QueryCommands.scala:
```diff
@@ -60,15 +60,16 @@ object QueryCommands {
     */
-  final case class QueryOptions(spreadProvider: SpreadProvider = StaticSpreadProvider(),
+     // to do see if can make spreadProvider as Options[Int] issue with QueryEngine
+  final case class QueryOptions(spread: Option[Int]= None,
```
spacing around `=`

velvia replied:
Actually do we really want to replace the spreadProvider?   I think it is better to keep this API no for flexibility?

TanviBhavsar replied:
@velvia Vish suggested that Spreadprovider should not be a part of QueryOptions. I changed it to spread as input and created SpreadProvider in QueryActor

vishramachandran replied:
We could, but the param would then have to be `Option[SpreadProvider]` instead since it is an override which by default wont be provided. I was originally in favor of simplicity since this is an override really. But I am ok with keeping it an `Option[SpreadProvider]` too - if the changes are straightforward.

TanviBhavsar replied:
@velvia  Is Option[Int] fine or should I change it to Option[SpreadProvider]?

velvia replied:
The main argument for `Option[SpreadProvider]` is flexibility in case a power user wants to use Akka API to provide their own.  However, whatever SpreadProvider is included basically would have to be in the FiloDB codebase when deployed.  It's for future extensibility.

TanviBhavsar replied:
> The main argument for `Option[SpreadProvider]` is flexibility in case a power user wants to use Akka API to provide their own. However, whatever SpreadProvider is included basically would have to be in the FiloDB codebase when deployed. It's for future extensibility.

Changed to Option[SpreadProvider]

velvia commented on coordinator/src/main/scala/filodb.coordinator/client/QueryCommands.scala:
```diff
@@ -109,7 +110,7 @@ object QueryCommands {
   final case class ExplainPlan2Query(dataset: DatasetRef,
                                      logicalPlan: LogicalPlan2,
-                                     queryOptions: QueryOptions = QueryOptions(),
+                                     queryOptions: QueryOptions =  QueryOptions(),
```
Extra space

velvia commented on coordinator/src/main/scala/filodb.coordinator/queryengine/Utils.scala:
```diff
@@ -75,8 +76,8 @@ object Utils extends StrictLogging {
               case Some(shardHash) => shardMap.queryShards(shardHash,
-                                                           options.spreadProvider.spreadFunc(filters).last.spread)
-              case None            => throw new IllegalArgumentException(s"Must specify filters for $shardCols")
+                spreadProvider.spreadFunc(filters).last.spread)
```
indentation.  Should indent to past =>

velvia commented on coordinator/src/test/scala/filodb.coordinator/queryengine2/QueryEngineSpec.scala:
```diff
@@ -133,7 +132,8 @@ class QueryEngineSpec extends FunSpec with Matchers {
 
     // materialized exec plan
-    val execPlan = engine.materialize(logicalPlan, QueryOptions(FunctionalSpreadProvider(spreadFunc)))
+    val execPlan = engine.materialize(logicalPlan, QueryOptions(),FunctionalSpreadProvider(spreadFunc))
```
Again space after ,

velvia commented on coordinator/src/test/scala/filodb.coordinator/queryengine2/QueryEngineSpec.scala:
```diff
@@ -148,7 +148,7 @@ class QueryEngineSpec extends FunSpec with Matchers {
       Seq(SpreadChange(0, 1), SpreadChange(25000000, 2)) // spread change time is in ms
     }
-    val execPlan = engine.materialize(lp, QueryOptions(FunctionalSpreadProvider(spread)))
+    val execPlan = engine.materialize(lp, QueryOptions(),FunctionalSpreadProvider(spread))
```
Space after ,

velvia commented on coordinator/src/test/scala/filodb.coordinator/queryengine2/QueryEngineSpec.scala:
```diff
@@ -157,7 +157,7 @@ class QueryEngineSpec extends FunSpec with Matchers {
       Seq(SpreadChange(0, 1), SpreadChange(35000000, 2))
     }
-    val execPlan = engine.materialize(lp, QueryOptions(FunctionalSpreadProvider(spread)))
+    val execPlan = engine.materialize(lp, QueryOptions(),FunctionalSpreadProvider(spread))
```
spacing

velvia commented on coordinator/src/test/scala/filodb.coordinator/queryengine2/QueryEngineSpec.scala:
```diff
@@ -167,7 +167,7 @@ class QueryEngineSpec extends FunSpec with Matchers {
       Seq(SpreadChange(0, 1), SpreadChange(25000000, 2))
     }
-    val execPlan = engine.materialize(lp, QueryOptions(FunctionalSpreadProvider(spread)))
+    val execPlan = engine.materialize(lp, QueryOptions(),FunctionalSpreadProvider(spread))
```
spacing

velvia commented on coordinator/src/test/scala/filodb.coordinator/queryengine2/QueryEngineSpec.scala:
```diff
@@ -180,7 +180,7 @@ class QueryEngineSpec extends FunSpec with Matchers {
       Seq(SpreadChange(0, 1), SpreadChange(35000000, 2))
     }
-    val execPlan = engine.materialize(lp, QueryOptions(FunctionalSpreadProvider(spread)))
+    val execPlan = engine.materialize(lp, QueryOptions(),FunctionalSpreadProvider(spread))
```
spacing

velvia commented on core/src/main/resources/filodb-defaults.conf:
```diff
@@ -25,7 +25,7 @@ filodb {
 
   # # of shards each application/metric is spread out to = 2^spread
-  default-spread = 1
+  spread.default = 1
```
Please add a large block comment explaining the format of the spread {... } with an example.  The defaults file should contain very very verbose explanations of each option.

TanviBhavsar replied:
Added description

velvia commented on coordinator/src/main/scala/filodb.coordinator/queryengine2/QueryEngine.scala:
```diff
@@ -128,39 +131,47 @@ class QueryEngine(dataset: Dataset,
-      case lp: SeriesKeysByFilters =>         materializeSeriesKeysByFilters(queryId, submitTime, options, lp)
-      case lp: ApplyMiscellaneousFunction =>  materializeApplyMiscellaneousFunction(queryId, submitTime, options, lp)
+      case lp: RawSeries => materializeRawSeries(queryId, submitTime, options, lp, spreadProvider)
+      case lp: RawChunkMeta => materializeRawChunkMeta(queryId, submitTime, options, lp,
```
This is MUCH less readable.  Please change back to previous alignment with `=>` lined up

vishramachandran commented on coordinator/src/main/scala/filodb.coordinator/QueryActor.scala:
```diff
@@ -55,6 +57,14 @@ final class QueryActor(memStore: MemStore,
+  var filodbSpreadMap = new mutable.HashMap[String, Int]()
+  config.getConfig("filodb.spread.override").entrySet().asScala.
+    foreach { x => filodbSpreadMap.put(x.getKey(), x.getValue().unwrapped().asInstanceOf[Int])}
+  val applicationShardKeyName = dataset.options.nonMetricShardColumns(0)
```
We need to think about how to have multiple keys in here. This requirement will come soon and better to address right away. Lets talk in person.

vishramachandran commented on http/src/test/scala/filodb/http/PrometheusApiRouteSpec.scala:
```diff
@@ -73,9 +73,62 @@ class PrometheusApiRouteSpec extends FunSpec with ScalatestRouteTest with AsyncT
+      contentType shouldEqual ContentTypes.`application/json`
+      val resp = responseAs[ExplainPlanResponse]
+      resp.status shouldEqual "success"
+      resp.debugInfo.filter(_.startsWith("--E~SelectRawPartitionsExec")).length shouldEqual 8
```
👍

vishramachandran commented on http/src/test/scala/filodb/http/PrometheusApiRouteSpec.scala:
```diff
@@ -73,9 +73,62 @@ class PrometheusApiRouteSpec extends FunSpec with ScalatestRouteTest with AsyncT
+        contentType shouldEqual ContentTypes.`application/json`
+        val resp = responseAs[ExplainPlanResponse]
+        resp.status shouldEqual "success"
+        resp.debugInfo.filter(_.startsWith("--E~SelectRawPartitionsExec")).length shouldEqual 2
```
👍

vishramachandran commented on http/src/test/scala/filodb/http/PrometheusApiRouteSpec.scala:
```diff
@@ -73,9 +73,62 @@ class PrometheusApiRouteSpec extends FunSpec with ScalatestRouteTest with AsyncT
+  it("should take spread override value from config for app") {
+    setupDataset()
+    // Give the coordinator nodes some time to get started
+    Thread sleep 5000
```
Move the sleep into setupDataset

velvia commented on coordinator/src/main/scala/filodb.coordinator/queryengine2/QueryEngine.scala:
```diff
@@ -133,24 +133,23 @@ class QueryEngine(dataset: Dataset,
-        spreadProvider)
-      case lp: ApplyMiscellaneousFunction => materializeApplyMiscellaneousFunction(queryId, submitTime, options, lp,
-        spreadProvider)
+                                              spreadProvider)
```
Sorry minor nit, could you also have the => line up as well?  Thanks  :)

TanviBhavsar replied:
Did space adjustment

velvia commented on cli/src/main/scala/filodb.cli/CliMain.scala:
```diff
@@ -60,6 +60,7 @@ class Arguments extends FieldArgs {
   var ignoreTagsOnPartitionKeyHash: Seq[String] = Nil
   var everyNSeconds: Option[String] = None
   var shards: Option[Seq[String]] = None
+  var spread: Option[String] = None
```
Yes it should work.

TanviBhavsar replied:
Changed to Option[Integer]. Option[Int]throws Exception in thread "main" com.quantifind.sumac.ArgException: Error parsing "3" into field "spread" (type = scala.Option<java.lang.Object>)

velvia commented on conf/timeseries-filodb-server.conf:
```diff
@@ -5,6 +5,13 @@ filodb {
+    default = 1
+    override {
+      App-0 = 2
+      App-5 = 2
```
With potential multiple key support I think a better config structure would be to have the partition key column/label name be in the config.  This also makes things much clearer, instead of an implicit first non-metric shard key.  So like this:

```
   override {
       _ns {
          App-0 = 2
       }
    }
```

TanviBhavsar replied:
Multiple worskpaces can also have same application/namespace name?
So should it be something like
```
override {
	ws1 {
		_ns {
			App - 0 = 2
		}
	},
	ws2 {
		_ns {
			App - 0 = 3
		}
	}
}
```

TanviBhavsar replied:
I have changed config for spread-override as discussed

TanviBhavsar commented:
> Great change overall.
> 
> 1. Please watch and fix for all indentation and spacing issues.  Better yet, try to tweak ScalaStyle so that it catches your style issues automatically.
> 2. Please document all new HTTP query options in `doc/http-api.md`
> 3. Please add docs in `doc/sharding.md` about spread, and configurable per-app spread options.
> 4. we need to have open source gateway parse the configurable spread properly, otherwise this change will result in spread incompatibility and data being lost for OSS users
> 
> @tjackpaul can we please have gateway take the spread options from the same config file format

Thanks Evan| I have made all the changes and updated docs. I have also added EQUALS and COMMA to org.scalastyle.scalariform.EnsureSingleSpaceAfterTokenChecker to catch space issues. Due to this I had to fix space in other files also.

TanviBhavsar commented on conf/timeseries-dev-source.conf:
```diff
@@ -1,5 +1,5 @@
     dataset = "prometheus"
-    num-shards = 4
+    num-shards = 8
```
Yes I have kept it as 4

TanviBhavsar commented:
> Nice set of elaborate changes and test cases!
> Query Engine expertise is growing very well. 👍

Thanks Vish! I have addressed your review comments

velvia commented on conf/timeseries-filodb-server.conf:
```diff
@@ -5,13 +5,19 @@ filodb {
+  spread-override = [
+    {
+      _ns = App-0,
+      spread = 2
```
Just one thought.  The design of config means that there cannot be a shard label named "spread", correct?  This should be documented.

TanviBhavsar replied:
Yes, I have added it in conf/timeseries-filodb-server.conf

TanviBhavsar replied:
That's a good idea. Cjanged it to _spread_

TanviBhavsar commented:
> @TanviBhavsar thanks for your work on this, great stuff.
> Please make sure the new spread override config is clearly documented in conf/timeseries-dev-source or wherever that config is.

Thanks Evan! I have documented spread-override in conf/timeseries-filodb-server.conf

vishramachandran commented on coordinator/src/main/scala/filodb.coordinator/QueryActor.scala:
```diff
@@ -91,11 +110,20 @@ final class QueryActor(memStore: MemStore,
+    val spreadProvider: SpreadProvider = if (queryOptions.spreadProvider.isDefined)
+      queryOptions.spreadProvider.get
+    else
+      functionalSpreadProvider
```
use option.map pattern

vishramachandran commented on conf/timeseries-filodb-server.conf:
```diff
@@ -5,6 +5,19 @@ filodb {
+  spread-default = 1
+
+  # Override default spread for application using override block which will have non metric shard keys and spread. There cannot be a shard label named "spread".
+  spread-override = [
```
shall we call this spread-assignment instead of "spread-override" ?
Different terms will help distinguish this from spread override passed in command line or query param in verbal conversations.

vishramachandran commented on coordinator/src/main/scala/filodb.coordinator/client/QueryCommands.scala:
```diff
@@ -44,6 +44,7 @@ object QueryCommands {
       Seq(spreadChange)
     }
   }
+  case class SpreadOverride(shardKeysMap: Map[String, String], spread: Int)
```
Rename to SpreadAssignment

vishramachandran commented on http/src/main/scala/filodb/http/PrometheusApiRoute.scala:
```diff
@@ -108,11 +108,16 @@ class PrometheusApiRoute(nodeCoord: ActorRef, settings: HttpSettings)(implicit a
+    val spreadProvider = if (spread.isDefined)
+      Some(StaticSpreadProvider(SpreadChange(0, spread.get)))
+    else
+      None
```
option.map pattern

TanviBhavsar commented:
> Approved, after some minor suggestions.
> Nice work! I am glad that the expanded scope covers some future use cases.

Thanks! I have made the changes
